dqvm/src/addresses.lisp

```diff
@@ -0,0 +1,334 @@
+  (when address
+    (with-slots (number-of-processes block-size blocks-per-process remainder-blocks total-blocks)
+        addresses
+
+      (assert (zerop (mod block-size 2)))
+
+      (let ((initial-address (get-initial-address addresses address)))
+        (if (< initial-address (* (- total-blocks remainder-blocks) block-size))
+            (nth-value 0 (floor initial-address (* blocks-per-process block-size)))
+            (mod (floor initial-address block-size) number-of-processes))))))
+
+(defmethod member-p ((addresses addresses) address)
+  "Return T if ADDRESS is in the table ADDRESSES."
+  (= (get-rank addresses address) (rank addresses)))
+
+(defmethod offset-of ((addresses addresses) address)
+  "Find the index (offset) of the amplitude corresponding to ADDRESS. Return NIL if address is not in ADDRESSES."
+  ;; XXX blend member-p with this function to avoid redundant work.
+  (when (member-p addresses address)
+    (with-slots (block-size blocks-per-process remainder-blocks total-blocks permutation rank)
+        addresses
+
+      (let* ((initial-address (get-initial-address addresses address))
+             (start-of-remainder-addresses (* (- total-blocks remainder-blocks) block-size)))
+
+        (if (< initial-address start-of-remainder-addresses)
+            (- initial-address (* rank blocks-per-process block-size))
+            (+ (* blocks-per-process block-size)
+               (- initial-address start-of-remainder-addresses (* rank block-size))))))))
+
+(defmethod %number-of-addresses ((addresses addresses))
+  "Number of addresses handled by the table ADDRESSES."
+  (multiple-value-bind (blocks-per-process remainder)
+      (floor (%get-total-blocks addresses)
+             (number-of-processes addresses))
+
+    (* (block-size addresses)
+       (+ blocks-per-process (if (< (rank addresses) remainder)
+                                 1
+                                 0)))))
+
+(defmethod effective-permutation ((addresses addresses) next-permutation)
+  "Return permutation required to map the original ADDRESSES to NEXT-PERMUTATION.
+
+If we have a table that is the image of (0, 1, ...) by the permutation πₖ and we want to map the original indices to πₕ, then we need to apply the effective permutation πₕ ∘ πₖ⁻¹."
+  (let ((curr-permutation (permutation addresses)))
+    (append (reverse curr-permutation) next-permutation)))
+
+(defmethod get-effective-permutation ((addresses addresses) next-permutation)
+  "Get the next permutation needed to apply PERMUTATION on ADDRESSES.
+
+Returns the effective permutation (i.e., π₂ ∘ π₁⁻¹) as well as the next permutation (i.e., π₂). These permutations are in a format ready to be passed to APPLY-QUBIT-PERMUTATION."
+  (let ((prev-permutation (permutation addresses)))
+    (append (reverse prev-permutation) next-permutation)))
+
+(defmethod update-permutation ((addresses addresses) next-permutation)
```

Comment:
just set `:writer update-permutation` in the class, and eliminate this method. It's ok if you have a reader/writer/accessor or even multiple of them. They'll all be methods.